crates/oxc_regular_expression/src/display.rs

```diff
@@ -0,0 +1,586 @@
+    F: Fn(&mut Peekable<I::IntoIter>) -> Option<String>,
+{
+    let sep = sep.as_ref();
+    let iter = &mut items.into_iter().peekable();
+
+    if let Some(first) = next(iter) {
+        write!(f, "{first}")?;
+    }
+
+    while let Some(it) = next(iter) {
+        write!(f, "{sep}{it}")?;
+    }
+
+    Ok(())
+}
+
+fn character_to_string(
+    this: &Character,
+    peek: Option<&Character>,
+) -> (/* result */ String, /* true of peek is consumed */ bool) {
+    const USED_PEEK: bool = true;
+    const UNUSED_PEEK: bool = false;
+
+    if is_trail_surrogate(this.value) {
+        return (format!(r"\u{:X}", this.value), UNUSED_PEEK);
+    }
+
+    let (cp, peeked) = if is_lead_surrogate(this.value) {
+        if peek.is_some_and(|peek| is_trail_surrogate(peek.value)) {
+            let Some(peek) = peek else { unreachable!() };
+            (combine_surrogate_pair(this.value, peek.value), USED_PEEK)
+        } else {
+            return (format!(r"\u{:X}", this.value), UNUSED_PEEK);
+        }
+    } else {
+        (this.value, UNUSED_PEEK)
+    };
+
+    // NOTE: even if `ch` is unused this conversion makes sure that we are looking at a correct value.
+    let Some(ch) = char::from_u32(cp) else { panic!("Invalid `Character` `{cp}`!") };
+    let result = match this.kind {
+        CharacterKind::ControlLetter => match ch {
+            '\n' => r"\cJ".to_string(),
+            '\r' => r"\cM".to_string(),
+            '\t' => r"\cI".to_string(),
+            _ => format!(r"\c{ch}"),
+        },
+        CharacterKind::Identifier => {
+            format!(r"\{ch}")
+        }
+        CharacterKind::Symbol => format!("{ch}"),
+        CharacterKind::Null => String::from(r"\0"),
+        CharacterKind::UnicodeEscape => {
+            // we remove the leading `0x` of our 4 digit hex number.
+            let hex = &format!("{cp:#4X}")[2..];
+            if hex.len() <= 4 {
+                format!(r"\u{hex}")
+            } else {
+                format!(r"\u{{{hex}}}")
+            }
+        }
+        CharacterKind::HexadecimalEscape => {
+            // we remove the leading `0x` of our 2 digit hex number.
+            let hex = &format!("{cp:#2X}")[2..];
+            format!(r"\x{hex}")
+        }
+        CharacterKind::Octal => {
+            let octal = format!("{cp:o}");
+            format!(r"\{octal}")
+        }
+        CharacterKind::SingleEscape => match ch {
+            '\n' => String::from(r"\n"),
+            '\r' => String::from(r"\r"),
+            '\t' => String::from(r"\t"),
+            '\u{b}' => String::from(r"\v"),
+            '\u{c}' => String::from(r"\f"),
+            '\u{8}' => String::from(r"\b"),
+            '\u{2D}' => String::from(r"\-"),
+            _ => format!(r"\{ch}"),
+        },
+    };
+
+    (result, peeked)
+}
+
+#[cfg(test)]
+mod test {
+    use oxc_allocator::Allocator;
+    use oxc_span::Span;
+
+    use crate::{literal_parser::parse_reg_exp_literal, FlagsParser, PatternParser};
```

Comment:
I also like to fix the following 2 points.

- `parse_reg_exp_literal` is being used like utils
  - It was not originally intended to be exposed to the outside
- The flags to test in unicode mode and the inclusion of the `u|v` flag in the literal seem as a double intent

So how about changing the test case just like this?

```rs
// Before
("/ab/", None, true),
(r"/\c_/", None, false),

// After
("/ab/", None),
("/ab/u", None),
(r"/\c_/", None),
```

then

```rs
    #[test]
    fn test_display() {
        let allocator = Allocator::default();
        for &(source, expect) in CASES {
            let expect = expect.unwrap_or(source);
            let actual = Parser::new(&allocator, source, ParserOptions::default()).parse().unwrap();
            assert_eq!(expect, actual.to_string());
        }
    }
```

Although somewhat verbose, it is more declarative and eliminates all of the above problems while also cleaning up the logic in the test code.